core/modules/layout/plugins/access/user_compare_layout_access.inc

```diff
@@ -0,0 +1,70 @@
+<?php
+/**
+ * @file
+ * Plugin to provide access if two user account contexts are equal.
+ */
+class UserCompareLayoutAccess extends LayoutAccess {
+  /**
+   * Constructor for a Layout access rule.
+   */
+  function __construct($plugin_name, array $data = array()) {
+    parent::__construct($plugin_name, $data);
+    $this->settings += array(
+      'equality' => 1,
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  function summary($contexts = array()) {
+    if (empty($this->contexts) || count($this->contexts) != 2 || empty($this->contexts['user1']->data) || empty($this->contexts['user2']->data)) {
+      return t('Comparing two User accounts');
+    }
+
+    // @todo for some reason context info is not available here.
+    $comparison = !empty($this->settings['equality']) ? "is" : 'is not';
+    return t('@id1 @comp @id2', array(
+      '@comp' => $comparison,
+      '@id1' => $this->contexts['user1']->data->uid,
+      '@id2' => $this->contexts['user2']->data->uid,
+    ));
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  function checkAccess() {
+    if (empty($this->contexts) || count($this->contexts) != 2 || empty($this->contexts['user1']->data) || empty($this->contexts['user2']->data)) {
+      return FALSE;
+    }
+
+    $account1 = $this->contexts['user1']->data;
+    $account2 = $this->contexts['user2']->data;
+
+    // xor returns false if the two bools are the same, and true if they are not.
+    // i.e, if we asked for equality and they are equal, return true.
+    // If we asked for INequality and they are equal, return false.
+    return ($account1->uid == $account2->uid) xor empty($this->settings['equality']);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  function form(&$form, &$form_state) {
+    parent::form($form, $form_state);
+    $form['helptext'] = array(
+      '#type' => 'markup',
+      '#value' => '<div>' . t('Compare the two user accounts. For example, for a user to view their own profile, choose "logged in user" and "user being viewed" and say "grant access if equal".') . '</div>',
```

Comment:
nit: we say "equal" here, but `$form['equality']['#options'][1]` below is using the word "same".

Reply:
addressed in the latest PR.